contracts/auctions/src/contract.rs

```diff
@@ -0,0 +1,455 @@
+            Some(current_highest_bid) if bid_amount > current_highest_bid => {
+                // refund the previous highest bidder
+                let old_bid_info = get_highest_bid(&env, auction_id)?;
+                token_client.transfer(
+                    &env.current_contract_address(),
+                    &old_bid_info.bidder,
+                    &(old_bid_info.bid as i128),
+                );
+            }
+            Some(_) => {
+                log!(
+                    &env,
+                    "Auction: Place Bid: Bid not enough. Amount bid: ",
+                    bid_amount
+                );
+                return Err(ContractError::BidNotEnough);
+            }
+            None => {}
+        };
+
+        token_client.transfer(
+            &bidder,
+            &env.current_contract_address(),
+            &(bid_amount as i128),
+        );
+
+        set_highest_bid(&env, auction_id, bid_amount, bidder.clone())?;
+
+        auction.highest_bid = Some(bid_amount);
+        save_auction(&env, &auction)?;
+
+        env.events()
+            .publish(("place bid", "auction id"), auction_id);
+        env.events().publish(("place bid", "bidder"), bidder);
+        env.events().publish(("place bid", "bid"), bid_amount);
+
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    pub fn finalize_auction(env: Env, auction_id: u64) -> Result<(), ContractError> {
+        let mut auction = get_auction_by_id(&env, auction_id)?;
+
+        // Check if the auction can be finalized
+        if auction.status != AuctionStatus::Active {
+            log!(
+                env,
+                "Auction: Finalize auction: Cannot finalize an inactive/ended auction."
+            );
+            return Err(ContractError::AuctionNotActive);
+        }
+        if env.ledger().timestamp() < auction.end_time {
+            log!(
+                env,
+                "Auction: Finalize auction: Auction cannot be ended early"
+            );
+            return Err(ContractError::AuctionNotFinished);
+        }
+
+        let token_client = token::Client::new(&env, &auction.auction_token);
+        let highest_bid = get_highest_bid(&env, auction_id)?;
+
+        // check if minimum price has been reached
+        if auction.item_info.minimum_price.map_or(true, |min_price| {
+            auction
+                .highest_bid
+                .map_or(false, |highest_bid| highest_bid >= min_price)
+        }) {
+            token_client.transfer(
+                &env.current_contract_address(),
+                &auction.seller,
+                &(highest_bid.bid as i128),
+            );
+
+            let nft_client = collection::Client::new(&env, &auction.item_info.collection_addr);
+            nft_client.safe_transfer_from(
+                &env.current_contract_address(),
+                &auction.seller,
+                &highest_bid.bidder,
+                &auction.item_info.item_id,
+                &1,
```

Comment:
I guess the amount of items should also be specified in the `ItemInfo` structure?

Reply:
moving this to a separate issue